File: Baekjoon/Class2/31.10816 숫자 카드 2.cpp
// 이진 탐색
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    cin >> n;
    vector<int> cards(n);
    for (int i = 0; i < n; i++)
        cin >> cards[i];
    sort(cards.begin(), cards.end()); // 정렬

    cin >> m;
    while (m--) {
        int target;
        cin >> target;
        // 시작과 끝 인덱스 이진 탐색 O(logn)
        auto target_range = equal_range(cards.begin(), cards.end(), target);
        cout << target_range.second - target_range.first << ' ';
        // auto target_lb = lower_bound(cards.begin(), cards.end(), target);
        // auto target_ub = upper_bound(cards.begin(), cards.end(), target);
        // cout << distance(target_lb, target_ub) << ' ';
    }
    return 0;
}

/* 이진 탐색 구현
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int bin_search_lower_bound(const vector<int>& vec, int target) {
    int lower_bound = -1;
    int left = 0, right = vec.size() - 1;
    while (left <= right) {
        int middle = (left + right) / 2;
        if (vec[middle] >= target) {
            if (vec[middle] == target) // 찾음
                lower_bound = middle;
            right = middle - 1; // 찾았어도 왼쪽 범위 계속 확인
        }
        else
            left = middle + 1;
    }
    return lower_bound;
}

int bin_search_upper_bound(const vector<int>& vec, int target) {
    int upper_bound = -1;
    int left = 0, right = vec.size() - 1;
    while (left <= right) {
        int middle = (left + right) / 2;
        if (vec[middle] <= target) {
            if (vec[middle] == target) // 찾음
                upper_bound = middle + 1;
            left = middle + 1; // 찾았어도 오른쪽 범위 계속 확인
        }
        else
            right = middle - 1;
    }
    return upper_bound;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;

    cin >> n;
    vector<int> cards(n);
    for (int i = 0; i < n; i++)
        cin >> cards[i];
    sort(cards.begin(), cards.end()); // 정렬

    cin >> m;
    for (int _ = 0; _ < m; _++) {
        int target;
        cin >> target;
        // 시작과 끝 인덱스 이진 탐색 O(logn)
        int target_lb = bin_search_lower_bound(cards, target);
        if (target_lb == -1) { // 시작 인덱스 없음
            cout << "0 ";
            continue;
        }
        int target_ub = bin_search_upper_bound(cards, target);
        cout << target_ub - target_lb << ' ';
    }
    return 0;
}
*/

/* unordered_map 사용
#include <iostream>
#include <unordered_map>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    unordered_map<int, int> cards(n);

    cin >> n;
    for (int i = 0; i < n; i++) {
        int card;
        cin >> card;
        cards[card]++;
    }

    cin >> m;
    while (m--) {
        int target;
        cin >> target;
        // unordered_map의 접근 시간 O(1)
        cout << cards[target] << ' ';
    }
    return 0;
}
*/

File: Baekjoon/Class2/31.10816 숫자 카드 2.java
// HashMap 사용
import java.io.*;
import java.util.HashMap;
import java.util.StringTokenizer;

class Main {
    public static void main(String[] args) throws IOException {
        int n, m;
        HashMap<Integer, Integer> cards = new HashMap<>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < n; i++) {
            int card = Integer.parseInt(st.nextToken());
            cards.put(card, cards.getOrDefault(card, 0) + 1); // 개수 세기
        }

        m = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < m; i++) {
            int target = Integer.parseInt(st.nextToken());
            // HashMap의 접근 시간 O(1)
            sb.append(cards.getOrDefault(target, 0)).append(' ');
        }
        System.out.print(sb);
        br.close();
    }
}

/* 이진 탐색 구현
import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {
    private static int binSearchLowerBound(int[] arr, int target) {
        int lowerBound = -1;
        int left = 0, right = arr.length - 1;
        while(left <= right) {
            int middle = (left + right) / 2;
            if(arr[middle] >= target) {
                if(arr[middle] == target) // 찾음
                    lowerBound = middle;
                right = middle - 1; // 찾았어도 왼쪽 범위 계속 확인
            }
            else
                left = middle + 1;
        }
        return lowerBound;
    }

    private static int binSearchUpperBound(int[] arr, int target) {
        int upperBound = -1;
        int left = 0, right = arr.length - 1;
        while(left <= right) {
            int middle = (left + right) / 2;
            if(arr[middle] <= target) {
                if(arr[middle] == target) // 찾음
                    upperBound = middle + 1;
                left = middle + 1; // 찾았어도 오른쪽 범위 계속 확인
            }
            else
                right = middle - 1;
        }
        return upperBound;
    }

    public static void main(String[] args) throws IOException {
        int n, m;
        int[] cards;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        cards = new int[n];
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < n; i++) {
            int card = Integer.parseInt(st.nextToken());
            cards[i] = card;
        }
        Arrays.sort(cards); // 정렬

        m = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < m; i++) {
            int target = Integer.parseInt(st.nextToken());
            // 시작과 끝 인덱스 이진 탐색 O(logn)
            int targetLB = binSearchLowerBound(cards, target);
            if(targetLB == -1) { // 시작 인덱스 없음
                sb.append("0 ");
                continue;
            }
            int targetUB = binSearchUpperBound(cards, target);
            sb.append(targetUB - targetLB).append(' ');
        }
        System.out.print(sb);
        br.close();
    }
}
*/